File: .tmp_sc4013/java_22/3923_minimum_generations_to_target_point.java
// CONFIG class=Solution method=minGenerations types=None
// LeetCode 3923 - Minimum Generations to Target Point
// https://leetcode.com/problems/minimum-generations-to-target-point/

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

class Solution {
    static class P {
        final int a, b, c;
        P(int a, int b, int c) { this.a = a; this.b = b; this.c = c; }
        @Override public boolean equals(Object o) {
            if (!(o instanceof P)) return false;
            P p = (P) o;
            return a == p.a && b == p.b && c == p.c;
        }
        @Override public int hashCode() { return Objects.hash(a, b, c); }
    }

    public int minGenerations(int[][] points, int[] target) {
        P targetPoint = new P(target[0], target[1], target[2]);
        Map<P, Integer> generation = new HashMap<>();
        List<P> all = new ArrayList<>();
        for (int[] values : points) {
            P p = new P(values[0], values[1], values[2]);
            generation.put(p, 0);
            all.add(p);
        }
        if (generation.containsKey(targetPoint)) return generation.get(targetPoint);
        for (int current = 1; ; current++) {
            int limit = all.size();
            List<P> added = new ArrayList<>();
            for (int i = 0; i < limit; i++) {
                for (int j = i + 1; j < limit; j++) {
                    if (all.get(i).equals(all.get(j))) continue;
                    P pi = all.get(i), pj = all.get(j);
                    P p = new P((pi.a + pj.a) / 2, (pi.b + pj.b) / 2, (pi.c + pj.c) / 2);
                    if (!generation.containsKey(p)) {
                        generation.put(p, current);
                        added.add(p);
                    }
                }
            }
            if (generation.containsKey(targetPoint)) return generation.get(targetPoint);
            if (added.isEmpty()) return -1;
            all.addAll(added);
        }
    }
}

File: .tmp_rs4013/rust_check22/3923_minimum_generations_to_target_point.rs
struct Solution;
// LeetCode 3923 - Minimum Generations to Target Point
// https://leetcode.com/problems/minimum-generations-to-target-point/

use std::collections::BTreeMap;

impl Solution {
    pub fn min_generations(points: Vec<Vec<i32>>, target: Vec<i32>) -> i32 {
        type Point = [i32; 3];
        let target_point: Point = [target[0], target[1], target[2]];
        let mut generation: BTreeMap<Point, i32> = BTreeMap::new();
        let mut all: Vec<Point> = Vec::new();
        for values in points {
            let p: Point = [values[0], values[1], values[2]];
            generation.insert(p, 0);
            all.push(p);
        }
        if generation.contains_key(&target_point) {
            return generation[&target_point];
        }
        let mut current = 1;
        loop {
            let limit = all.len();
            let mut added = Vec::new();
            for i in 0..limit {
                for j in (i + 1)..limit {
                    if all[i] == all[j] {
                        continue;
                    }
                    let p: Point = [
                        (all[i][0] + all[j][0]) / 2,
                        (all[i][1] + all[j][1]) / 2,
                        (all[i][2] + all[j][2]) / 2,
                    ];
                    if !generation.contains_key(&p) {
                        generation.insert(p, current);
                        added.push(p);
                    }
                }
            }
            if generation.contains_key(&target_point) {
                return generation[&target_point];
            }
            if added.is_empty() {
                return -1;
            }
            all.extend(added);
            current += 1;
        }
    }
}

fn main() {}
